File: MinPathFallingSum.java
package DP3;

//Time complexity : O(n^2) as we are calculating min for each element
//Space complexity: O(1)
class MinPathFallingSum {
    public int minFallingPathSum(int[][] matrix) {
        if(matrix == null || matrix.length == 0)
            return -1;
        int min = Integer.MAX_VALUE;
        for(int i=matrix.length-2; i>=0; i--){
            for(int j=0;j<matrix.length; j++){
                int a = Integer.MAX_VALUE, b = Integer.MAX_VALUE, c = Integer.MAX_VALUE;
                if(j == 0 && j == matrix.length-1){
                    b = matrix[i+1][j];
                } else if(j == 0){
                    b = matrix[i+1][j];
                    c = matrix[i+1][j+1];
                } else if(j == matrix.length -1){
                    a = matrix[i+1][j-1];
                    b = matrix[i+1][j];
                } else{
                    a = matrix[i+1][j-1];
                    b = matrix[i+1][j];
                    c = matrix[i+1][j+1];
                }
                matrix[i][j] = matrix[i][j] + Math.min(a, Math.min(b, c));
            }
        }
        for(int j=0; j< matrix.length; j++){
            min = Math.min(min, matrix[0][j]);
        }
        return min;
    }
}

File: minPathFallingSum.py
#Time Complexity: O(row*col)
#Space Complexity: O(row*col)
#Did it run on leetcode: Yes

class Solution:
    def minFallingPathSum(self, matrix: List[List[int]]) -> int:
        rows = len(matrix)
        columns = len(matrix[0])
        dp = [[0]*columns for index in range(rows)]
        for jindex in range(columns):
            dp[0][jindex] = matrix[0][jindex]
        for index in range(1, rows):
            for jindex in range(columns):
                if jindex == 0:
                    dp[index][jindex] = min(dp[index - 1][jindex] + matrix[index][jindex],
                                            dp[index - 1][jindex + 1] + matrix[index][jindex])
                elif jindex == columns - 1:
                    dp[index][jindex] = min(dp[index - 1][jindex] + matrix[index][jindex],
                                           dp[index - 1][jindex - 1] + matrix[index][jindex])
                else:
                    dp[index][jindex] = min(dp[index - 1][jindex - 1] + matrix[index][jindex],
                                           dp[index - 1][jindex] + matrix[index][jindex],
                                           dp[index - 1][jindex + 1] + matrix[index][jindex])
        return min(dp[-1])